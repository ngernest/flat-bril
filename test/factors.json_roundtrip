{
  "functions": [
    {
      "args": [
        {
          "name": "num",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "zer",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "dest": "fac",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "label": "loopcheck"
        },
        {
          "args": [
            "one",
            "num"
          ],
          "dest": "ispos",
          "funcs": [],
          "labels": [],
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "ispos"
          ],
          "funcs": [],
          "labels": [
            "loopbody",
            "loopend"
          ],
          "op": "br"
        },
        {
          "label": "loopbody"
        },
        {
          "args": [
            "num",
            "fac"
          ],
          "dest": "quo",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "quo",
            "fac"
          ],
          "dest": "tmp",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "num",
            "tmp"
          ],
          "dest": "mod",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "mod",
            "zer"
          ],
          "dest": "iszero",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "iszero"
          ],
          "funcs": [],
          "labels": [
            "ifyes",
            "ifno"
          ],
          "op": "br"
        },
        {
          "label": "ifyes"
        },
        {
          "args": [
            "fac"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "args": [
            "num",
            "fac"
          ],
          "dest": "num",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "loopcheck"
          ],
          "op": "jmp"
        },
        {
          "label": "ifno"
        },
        {
          "args": [
            "fac",
            "one"
          ],
          "dest": "fac",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "loopcheck"
          ],
          "op": "jmp"
        },
        {
          "label": "loopend"
        }
      ],
      "name": "main"
    }
  ]
}
