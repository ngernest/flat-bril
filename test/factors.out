@main
	op: const		dest: zer		type: int		value: 0	
	op: const		dest: one		type: int		value: 1	
	op: const		dest: fac		type: int		value: 2	
.loopcheck
	op: lt   		dest: ispos		type: bool		args: ["one", "num"]	
	op: br   		args: ["ispos"]		labels: ["loopbody", "loopend"]
.loopbody
	op: div  		dest: quo		type: int		args: ["num", "fac"]	
	op: mul  		dest: tmp		type: int		args: ["quo", "fac"]	
	op: sub  		dest: mod		type: int		args: ["num", "tmp"]	
	op: eq   		dest: iszero		type: bool		args: ["mod", "zer"]	
	op: br   		args: ["iszero"]		labels: ["ifyes", "ifno"]
.ifyes
	op: print		args: ["fac"]	
	op: div  		dest: num		type: int		args: ["num", "fac"]	
	op: jmp  		labels: ["loopcheck"]
.ifno
	op: add  		dest: fac		type: int		args: ["fac", "one"]	
	op: jmp  		labels: ["loopcheck"]
.loopend
